### PR TITLE
feat(skills): add vss-run-gym-eval for scoring a deployment with NeMo Gym

### PR DESCRIPTION
**Pull request by simings-nv**
## Description

Adds `vss-run-gym-eval`, a developer-side skill for evaluating a running VSS deployment with **NVIDIA NeMo Gym**. An evaluation becomes a scalar reward per task — the contract a training loop consumes — and one deployment can be scored by both NeMo Gym and VSS's own harness for comparison.

### Documentation and eval specs only

`SKILL.md`, three references, three eval specs, their Harbor adapter, a skill card, and two catalog rows in `skills/README.md`.

**No third-party dependency is introduced** — no requirements file, no pyproject, no imports. Scoring is performed by a NeMo Gym *resources server* that is **not vendored here**: it implements Gym's interface and imports Gym, which VSS never runs. The skill documents how to obtain and stage it.

**No shipped file changes.** No developer profile, no service in any compose file, no entry in `containers.env` or `container-inventory.json`. The eval runner is composed at runtime as a Delta Profile on an existing developer profile, landing in `_builds/<name>/` — gitignored, and per `.gitignore:28-32` never a Compose profile. Nothing runs until the skill is invoked.

## Using it

Ask a coding agent to evaluate a deployment — *"evaluate this deployment with Gym"*, *"run the Gym eval"*, *"compare our eval against Gym's"* — and the skill:

1. **Checks the image is safe to pull.** The published Gym image predates the upstream removal of royalty-bearing codec libraries, so the gate establishes build provenance from the registry manifest list, platform manifest and config blob — roughly 20 KB, no image layers — and **exits non-zero** unless the build postdates the fix and no layer records a codec install.
2. **Composes the runner as a delta** on the profile already deployed, adding one service key and nothing else.
3. **Runs the two-phase Gym lifecycle** — `gym env start`, then `gym eval run` — with the policy pointed at the deployment's own LLM endpoint, so no external provider is required.
4. **Returns a scalar reward per task**, with the judge rationale and per-check signals persisted for diagnosis.

**Scope.** This covers the ask-video path, where VSS is the environment and the policy queries it. It does not cover the CI skill-eval corpus — those tasks are shell execution in a sandbox and require Gym's `harbor_agent`, which pins a Harbor release incompatible with the `harbor==0.20.0` at `.github/skill-eval/run_leg.py:53`. Tracked upstream in [NVIDIA-NeMo/Gym#2596](https://github.com/NVIDIA-NeMo/Gym/pull/2596).

## Two things the documentation is deliberate about

**This is a distinct scorer, not a pass-through of VSS's own eval.** The server applies its own judge and its own formula — `0.4 × (the policy called the tool) + 0.6 × quality`. Its numbers are **not** numerically comparable to CI skill-eval scores; compare trends and failure modes, not absolute values.

**Only the host workflow is documented as verified**, because only it has been run end to end. `references/delta.md` marks the containerised runner as the packaging target with its four known gaps listed, and it cannot be exercised until a codec-clean image exists — running it means pulling an image the gate rejects.

## Validation

**End to end against a live VSS deployment**, following the documented steps from a clean stage:

```
policy calls ask_vss -> VSS answers from warehouse_safety_0001 -> LLM judge grades it
mean/reward 1.0   verifier_ok True   verifier_status "rubric"
```

`warehouse_safety_0001` is the clip four existing eval specs already use, so a Gym reward and VSS's own harness measure the same video.

**The image gate**, verified by exit status rather than output:

| Input | Result |
|---|---|
| clean build, postdates the fix | `exit 0` — GATE PASS |
| codec-installing layer | `exit 1` |
| clean but predates the fix | `exit 1` |
| the real published tag's metadata | `exit 1` |
| empty / whitespace / malformed / history-less config blob | `exit 1` |

Build dates are compared as epochs rather than strings, so an offset timestamp such as `2026-08-11T01:00:00+10:00` — which is `2026-08-10` in UTC, before the fix — is correctly rejected.

**Other checks**

- `skills/` plus the eval adapter under `.github/skill-eval/adapters/`
- Frontmatter parses with `name`, `description`, `license`, `metadata.version`; the name satisfies the playbook naming policy
- Eval corpus, measured develop vs. branch: **50 → 52 specs, 154 → 156 cells, 853 → 869 checks**; both new specs are `ANY` / `gpu_count: 0` and consume no GPU

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA-AI-Blueprints/video-search-and-summarization/blob/HEAD/CONTRIBUTING.md).
- [x] I have installed and run [pre-commit hooks](https://github.com/NVIDIA-AI-Blueprints/video-search-and-summarization/blob/HEAD/CONTRIBUTING.md#1-enable-pre-commit-hooks) locally.
- [x] Every commit on this PR is [DCO sign-off](https://github.com/NVIDIA-AI-Blueprints/video-search-and-summarization/blob/HEAD/CONTRIBUTING.md#developer-certificate-of-origin-dco)'d.
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by simings-nv** [comment]
@greptileai review

**Comment by simings-nv** [comment]
@greptileai review

File: skills/vss-run-gym-eval/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+without pulling the 13 GB image:
+
+```bash
+TOK=$(curl -s "https://nvcr.io/proxy_auth?scope=repository:nvidia/eval-factory/nemo-gym:pull" | jq -r .token)
+DIGEST=$(curl -s -H "Authorization: Bearer $TOK" \
+  -H "Accept: application/vnd.docker.distribution.manifest.list.v2+json" \
+  "https://nvcr.io/v2/nvidia/eval-factory/nemo-gym/manifests/${VSS_GYM_EVAL_TAG}" \
+  | jq -r '.manifests[] | select(.platform.architecture=="amd64") | .digest')
+# Inspect the config blob's creation date and layer history for codec installs.
+```
```
[the remark this answers is not in the document]
Comment (simings-nv) [reply skills/vss-run-gym-eval/SKILL.md:179]: Fixed in 6be209b12 — thank you, this was the finding that mattered most, since the gate was the only barrier before pulling a royalty-bearing image and as written it could not do its job.

It now walks all three hops: manifest list → platform manifest → **config blob**, and states the acceptance rule explicitly — `created` after 2026-08-11 **and** zero layers whose `created_by` installs ffmpeg/libav/x264/x265. Both conditions, because a later build date alone does not prove the codec removal is present.

Verified by running it against `26.05`:

```
created: 2026-06-01T12:53:28-07:00
codec-installing layers: 1
```

A clear fail, which is the gate working. About 20 KB is fetched; no layer is pulled.

File: skills/vss-eval-with-gym/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
+---
+name: vss-eval-with-gym
```
[the remark this answers is not in the document]
Comment (simings-nv) [reply skills/vss-eval-with-gym/SKILL.md]: Fixed in 6be209b12 — `skills/README.md:101` now carries:

```
| Score a deployment with NeMo Gym, or compare that against VSS's own eval | [`vss-eval-with-gym`](vss-eval-with-gym/SKILL.md) |
```

Agreed on the reasoning: a skill absent from that table is one an agent may never route to, which makes it effectively invisible however good its SKILL.md is.

(Both inline threads here predate that commit — your summary review at head `8b4e0c03c` already records both as resolved.)

**Comment by simings-nv** [comment]
@greptileai review

**Comment by simings-nv** [comment]
@greptileai review

**Comment by simings-nv** [comment]
@greptileai review

**Comment by simings-nv** [comment]
@greptileai review

**Comment by simings-nv** [comment]
@greptileai review

**Review by zac-wang-nv — CHANGES_REQUESTED** [review]
_This is Hinton, the VSS CI OSRB reviewer_

## Automated OSRB Review — GAP HARD

License Diff reports nemo-gym added (Apache Software, unpinned) with no OSRB approval source anywhere on bug 6111789; source app.py imports nemo_gym, fastapi, pydantic, httpx which the approved set does not cover. New third-party component/import additions require review and block.

- Dependency changes evaluated: `1`
- Public evidence: [License Diff Actions run](https://github.com/NVIDIA-AI-Blueprints/video-search-and-summarization/actions/runs/32216035522)
- Internal approval evidence was evaluated by the protected compliance pipeline.

| Package | Version | License | Classification |
|---|---|---|---|
| nemo-gym | latest (0.5.0) | Apache Software | NOT_APPROVED |

File: skills/vss-eval-with-gym/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+# 2. that manifest -> its config blob digest
+CFG=$(curl -fsS -H "Authorization: Bearer $TOK" \
+  -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.oci.image.manifest.v1+json" \
+  "https://nvcr.io/v2/${REPO}/manifests/${AMD}" | jq -r '.config.digest')
+[ -n "$CFG" ] || { echo "GATE FAIL: could not resolve the config blob -- do not proceed"; exit 1; }
+
+# 3. the config blob -> build date and layer history
+BLOB=$(curl -fsSL -H "Authorization: Bearer $TOK" "https://nvcr.io/v2/${REPO}/blobs/${CFG}")
+echo "created: $(echo "$BLOB" | jq -r '.created')"
+echo "codec-installing layers: $(echo "$BLOB" | jq -r '.history[].created_by' | grep -cE 'ffmpeg|libav|x264|x265')"
```
[the remark this answers is not in the document]
Comment (simings-nv) [reply skills/vss-eval-with-gym/SKILL.md]: Fixed. The config-blob hop now fails closed at every step:

```bash
BLOB=$(curl -fsSL ... ) || { echo "GATE FAIL: config blob fetch failed"; exit 1; }
[ -n "$BLOB" ] || { echo "GATE FAIL: empty config blob"; exit 1; }
CREATED=$(echo "$BLOB" | jq -er '.created') || { echo "GATE FAIL: no .created"; exit 1; }
[ -n "$CREATED" ] || { echo "GATE FAIL: empty .created"; exit 1; }
echo "$BLOB" | jq -e '.history | arrays and (length > 0)' >/dev/null || { echo "GATE FAIL: no usable .history"; exit 1; }
```

The emptiness test is not redundant, and my first attempt at this fix was still fail-open: `jq -er` on **empty** input produces no output and exits 0, so an empty blob sailed past the `.created` check. Caught by running the case rather than reading the patch.

Verified under `set -euo pipefail` — empty blob, whitespace-only, valid-JSON-without-history and non-JSON all reject; the real image reports its codec layer.

File: skills/README.md [hunk 4]
```diff
@@ -98,6 +98,7 @@ Match the user's intent to a skill. Start here before opening any individual `SK
 | Calibrate a multi-camera dataset (often a prerequisite for 3D) | [`vss-generate-video-calibration`](vss-generate-video-calibration/SKILL.md) |
 | Deploy behavior analytics on its own | [`vss-setup-behavior-analytics`](vss-setup-behavior-analytics/SKILL.md) |
 | Deploy the video-analytics REST API on its own | [`vss-setup-video-analytics-api`](vss-setup-video-analytics-api/SKILL.md) |
+| Score a deployment with NeMo Gym, or compare that against VSS's own eval | [`vss-eval-with-gym`](vss-eval-with-gym/SKILL.md) |
```
[the remark this answers is not in the document]
Comment (simings-nv) [reply skills/README.md]: Fixed. You were right that these are two different lists, and my earlier fix only touched one of them — the routing table under "Which skill do I need?", not the "Catalog (by layer)" section the install prompt reads.

Both now carry the skill. Since the install flow says "for each skill in the catalog", a skill missing from that section is silently skipped at install time however good its routing entry is.

File: skills/vss-eval-with-gym/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+# and exits 0, so an empty blob would sail past the `.created` check below. That
+# was verified, not assumed.
+[ -n "$BLOB" ] || { echo "GATE FAIL: empty config blob -- do not proceed"; exit 1; }
+CREATED=$(echo "$BLOB" | jq -er '.created') \
+  || { echo "GATE FAIL: no .created in config blob -- do not proceed"; exit 1; }
+[ -n "$CREATED" ] || { echo "GATE FAIL: empty .created -- do not proceed"; exit 1; }
+echo "$BLOB" | jq -e '.history | arrays and (length > 0)' >/dev/null \
+  || { echo "GATE FAIL: config blob has no usable .history -- do not proceed"; exit 1; }
+
+CODEC_LAYERS=$(echo "$BLOB" | jq -r '.history[].created_by // empty' | grep -cE 'ffmpeg|libav|x264|x265')
```
[the remark this answers is not in the document]
Comment (simings-nv) [reply skills/vss-eval-with-gym/SKILL.md]: Fixed, with your suggestion. Reproduced first:

```
$ bash -c 'set -e; C=$(echo "clean layer" | grep -cE "ffmpeg|libav|x264|x265"); echo reached'
exit=1     # never reached
```

One note for the record: `|| true` is correct here but `|| echo 0` is not, which I tried first. `grep -c` **prints 0 and exits 1** on no match, so the echo appends a second zero and `CODEC_LAYERS` becomes "0\n0", breaking the numeric comparison that follows.

Verified under `set -euo pipefail` across seven cases; a clean image now reaches ACCEPT instead of aborting.

File: skills/vss-run-gym-eval/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+  || { echo "GATE FAIL: config blob has no usable .history -- do not proceed"; exit 1; }
+
+# `|| true` is required, not defensive: grep -c exits 1 when it matches nothing,
+# so under `set -e` a CODEC-FREE image -- the one this gate exists to approve --
+# would abort the script before it could pass. Use `|| true`, NOT `|| echo 0`:
+# grep -c already prints 0 on no match, so echoing another produces "0\n0" and
+# breaks the numeric comparison below.
+CODEC_LAYERS=$(echo "$BLOB" | jq -r '.history[].created_by // empty' | grep -cE 'ffmpeg|libav|x264|x265' || true)
+echo "created: ${CREATED}"
+echo "codec-installing layers: ${CODEC_LAYERS}"
```
[the remark this answers is not in the document]
Comment (simings-nv) [reply skills/vss-run-gym-eval/SKILL.md:154]: Fixed — this was the right call and the most important of the gate findings. Printing two values is not a gate: a caller checking the exit status read success for exactly the image the gate exists to stop.

The script now decides, and exits non-zero with a specific reason:

```bash
FIX_EPOCH=$(date -u -d '2026-08-11' +%s)
CREATED_EPOCH=$(date -u -d "${CREATED}" +%s) || { echo "GATE FAIL: unparseable .created"; exit 1; }
[ "${CREATED_EPOCH}" -gt "${FIX_EPOCH}" ] || { echo "GATE FAIL: build predates the fix"; exit 1; }
[ "${CODEC_LAYERS}" -eq 0 ] || { echo "GATE FAIL: ${CODEC_LAYERS} layer(s) install codec packages"; exit 1; }
echo "GATE PASS"
```

Verified by exit status rather than output: clean + post-fix → 0; codec-bearing, pre-fix, the real 26.05 metadata, empty blob and history-less blob → 1.

One deliberate difference from the suggested form. `[[ "$CREATED" > "2026-08-11" ]]` compares strings, which is right for plain UTC stamps but wrong when the registry returns an offset:

```
created = 2026-08-11T01:00:00+10:00   (= 2026-08-10T15:00Z, i.e. BEFORE the fix)
  string comparison -> ACCEPT   (wrong)
  epoch comparison  -> REJECT   (correct)
```

Epoch parsing also handles the real value, which carries nanoseconds and an offset: `2026-06-01T12:53:28.624508933-07:00`.

**Comment by simings-nv** [comment]
@greptileai review

**Comment by simings-nv** [comment]
@greptileai review

Five commits since the last pass, addressing the Codex review. Summary of what changed:

- **Image gate** — rejects an image whose layer history cannot be inspected (`history: [{}]` previously passed), and rejects builds stamped on the cutoff day itself.
- **Eval specs** — the image-gate spec could be satisfied by reciting SKILL.md without contacting the registry; it now requires the fetch in the trajectory. The delta spec now asserts all four build artifacts, that the composed file parses, and that the Foundation-to-delta service diff is exactly one key.
- **Delta spec was unsatisfiable** — composition resolves the runner image through a fail-closed `\${VSS_GYM_EVAL_TAG:?}` with no default, and the task never supplied one. It now does, and a check asserts the fail-closed behaviour directly.
- **Adapter** — gold solutions replaced (both were `echo`-only, so the oracle could not satisfy its own checks); checks now address the checkout by absolute path rather than relative to an arbitrary cwd; `vss-build-vision-agent` bundled, since `delta.md` links into it.
- **Docs** — claims that stated unverified behaviour as fact are qualified, and the data-directory teardown warning corrected against `blueprint-deploy.sh`.

Both gold solutions were executed rather than reasoned about: the delta one composes and its service diff is exactly `> gym-eval`; the gate one fetches the live config blob for `26.05` and rejects it.

**Comment by simings-nv** [comment]
Superseded by a rebuilt branch. Closing this one.
